src/compiler/checker.ts

```diff
@@ -19901,6 +19903,28 @@ namespace ts {
             return isArrayType(type) || !(type.flags & TypeFlags.Nullable) && isTypeAssignableTo(type, anyReadonlyArrayType);
         }
 
+        function getSingleBaseForNonAugmentingSubtypeOrType(type: Type) {
+            if (!(getObjectFlags(type) & ObjectFlags.Reference) || !(getObjectFlags((type as TypeReference).target) & ObjectFlags.ClassOrInterface)) {
+                return type;
+            }
+            if ((type as TypeReference).cachedEquivalentBaseTypeOrSelf) {

```

Comment:
Looks like we take close to a 1% perf hit from adding this property to every type reference. Maybe use a pair of `ObjectFlags` to track whether (a) you've computed this information and (b) whether there is an equivalent base type, and only add the property then. You could co-opt the `IsNeverIntersectionComputed` and `IsNeverIntersection` bit positions for this use in type references as they're currently used only in intersections.

BTW, I also notice that someone added an `ObjectFlags.IsClassInstanceClone` with the value `1 << 30`. That's outside the small integer range and means the VM will resort to boxing and de-optimizing whenever that flags is set.

Reply:
I just put up #43732 which reorganizes the `ObjectFlags` enum. Should be pretty easy now to add the flags I suggest above.

Reply:
Done; but it looks like my comment on https://github.com/microsoft/TypeScript/pull/43732#pullrequestreview-639191529 was warranted, though - I had to fix a crash reusing those flags triggered. Should we audit our other `ObjectFlags` usages of these type-dependent flags to ensure there aren't other crashes that simply aren't triggered by our test suite?

Reply:
I did a quick audit and found one more location we were missing a `TypeFlags` check - I've opened https://github.com/microsoft/TypeScript/pull/43745 with just those changes.